File: app/src/main/java/by/anegin/telegram_contests/core/ui/model/UiChart.java
package by.anegin.telegram_contests.core.ui.model;

import android.graphics.Matrix;
import by.anegin.telegram_contests.core.data.model.Chart;
import by.anegin.telegram_contests.core.data.model.Column;

import java.text.SimpleDateFormat;
import java.util.ArrayList;
import java.util.Date;
import java.util.List;
import java.util.Locale;

public class UiChart {

    public final long[] xValues;
    public final List<UiDate> dates;
    public final List<Graph> graphs;
    public final long width;
    public final long minX;
    public final long minY;

    public UiChart(Chart chart) {
        if (chart.x.values.length == 0 || chart.lines.isEmpty()) {
            xValues = new long[0];
            dates = new ArrayList<>();
            graphs = new ArrayList<>();
            width = 0;
            minX = 0;
            minY = 0;
            return;
        }

        long[] minMaxX = findMinMax(chart.x.values);
        minX = minMaxX[0];

        xValues = new long[chart.x.values.length];
        for (int i = 0; i < xValues.length; i++) {
            xValues[i] = chart.x.values[i] - minMaxX[0];
        }

        long minY = Long.MAX_VALUE;
        long maxY = Long.MIN_VALUE;
        for (Column.Line line : chart.lines) {
            int count = Math.min(chart.x.values.length, line.values.length);
            for (int i = 0; i < count; i++) {
                if (line.values[i] > maxY) maxY = line.values[i];
                if (line.values[i] < minY) minY = line.values[i];
            }
        }

        this.minY = minY;

        Matrix matrix = new Matrix();
        matrix.setTranslate(-minMaxX[0], -minY);

        graphs = new ArrayList<>();
        for (Column.Line line : chart.lines) {
            float[] points = lineToPoints(chart.x.values, line.values);
            if (points.length > 0) {
                matrix.mapPoints(points);
                graphs.add(new Graph(line.id, line.values, points, line.color));
            }
        }

        SimpleDateFormat sdf = new SimpleDateFormat("MMM dd", Locale.US);
        dates = new ArrayList<>(chart.x.values.length);
        for (long xv : chart.x.values) {
            String dateString = sdf.format(new Date(xv));
            dates.add(new UiDate(xv - minMaxX[0], dateString));
        }

        width = minMaxX[1] - minMaxX[0];
    }

    private long[] findMinMax(long[] values) {
        long[] minMax = new long[]{values[0], values[0]};
        for (int i = 1; i < values.length; i++) {
            if (values[i] < minMax[0]) minMax[0] = values[i];
            if (values[i] > minMax[1]) minMax[1] = values[i];
        }
        return minMax;
    }

    private float[] lineToPoints(long[] x, long[] y) {
        int count = Math.min(x.length, y.length);
        if (count < 2) return new float[]{};

        float[] points = new float[4 * (count - 1)];
        points[0] = x[0];
        points[1] = y[0];
        int j = 2;
        for (int i = 1; i < count - 1; i++) {
            points[j] = x[i];
            points[j + 1] = y[i];
            points[j + 2] = points[j];
            points[j + 3] = points[j + 1];
            j += 4;
        }
        points[j] = x[count - 1];
        points[j + 1] = y[count - 1];

        return points;
    }

}

File: app/src/main/java/by/anegin/telegram_contests/core/ui/model/UiChart.kt
package by.anegin.telegram_contests.core.ui.model

import android.graphics.Matrix
import by.anegin.telegram_contests.core.data.model.Chart

class UiChart(
    val graphs: List<Graph>,
    val width: Long,
    val height: Long
) {

    companion object {
        fun fromChart(chart: Chart): UiChart? {
            if (chart.x.values.isEmpty() || chart.lines.isEmpty()) return null

            val xValuesCount = chart.x.values.size

            var minX = chart.x.values[0]
            var maxX = minX
            for (i in 1 until xValuesCount) {
                if (chart.x.values[i] > maxX) maxX = chart.x.values[i]
                if (chart.x.values[i] < minX) minX = chart.x.values[i]
            }

            var minY = Long.MAX_VALUE
            var maxY = Long.MIN_VALUE
            chart.lines.forEach { line ->
                val count = Math.min(xValuesCount, line.values.size)
                for (i in 0 until count) {
                    if (line.values[i] > maxY) maxY = line.values[i]
                    if (line.values[i] < minY) minY = line.values[i]
                }
            }

            val matrix = Matrix()
            matrix.setTranslate(-minX.toFloat(), -minY.toFloat())
            val graphs = chart.lines.mapNotNull { line ->
                val count = Math.min(xValuesCount, line.values.size)
                if (count > 1) {
                    val points = FloatArray(4 * (count - 1))
                    points[0] = chart.x.values[0].toFloat()
                    points[1] = line.values[0].toFloat()
                    var j = 2
                    for (i in 1 until count - 1) {
                        points[j] = chart.x.values[i].toFloat()
                        points[j + 1] = line.values[i].toFloat()
                        points[j + 2] = points[j]
                        points[j + 3] = points[j + 1]
                        j += 4
                    }
                    points[j] = chart.x.values[count - 1].toFloat()
                    points[j + 1] = line.values[count - 1].toFloat()
                    matrix.mapPoints(points)
                    Graph(points, line.color)
                } else {
                    null
                }
            }

            val chartWidth = maxX - minX
            val chartHeight = maxY - minY

            return UiChart(graphs, chartWidth, chartHeight)
        }
    }

}